File: src/ml_release_platform/domain/policy.py
"""Deterministic policy parsing and metric-gate evaluation."""

from __future__ import annotations

from collections.abc import Mapping
from dataclasses import dataclass
from math import isfinite
from typing import Any, Literal

from ml_release_platform.domain.errors import PolicyValidationError

GateOperator = Literal["min", "max"]


@dataclass(frozen=True, slots=True)
class PolicyGate:
    metric: str
    operator: GateOperator
    threshold: float


@dataclass(frozen=True, slots=True)
class ReleasePolicy:
    gates: tuple[PolicyGate, ...]


@dataclass(frozen=True, slots=True)
class GateEvaluation:
    metric: str
    operator: GateOperator
    threshold: float
    actual: float | None
    passed: bool
    reason: str | None = None


@dataclass(frozen=True, slots=True)
class PolicyEvaluation:
    passed: bool
    results: tuple[GateEvaluation, ...]


def _as_number(value: Any, field_name: str) -> float:
    if isinstance(value, bool) or not isinstance(value, int | float):
        raise PolicyValidationError(f"{field_name} must be a number.")
    numeric_value = float(value)
    if not isfinite(numeric_value):
        raise PolicyValidationError(f"{field_name} must be finite.")
    return numeric_value


def parse_policy(raw_policy: Mapping[str, Any]) -> ReleasePolicy:
    """Validate an external policy representation and return its deterministic form."""
    if not isinstance(raw_policy, Mapping):
        raise PolicyValidationError("Policy must be an object.")
    if set(raw_policy) != {"gates"}:
        raise PolicyValidationError("Policy must contain exactly one 'gates' field.")

    raw_gates = raw_policy["gates"]
    if not isinstance(raw_gates, Mapping) or not raw_gates:
        raise PolicyValidationError("Policy gates must be a non-empty object.")

    gates: list[PolicyGate] = []
    for metric, raw_gate in raw_gates.items():
        if not isinstance(metric, str) or not metric.strip():
            raise PolicyValidationError("Policy gate names must be non-empty strings.")
        if not isinstance(raw_gate, Mapping):
            raise PolicyValidationError(f"Gate '{metric}' must be an object.")
        if set(raw_gate) not in ({"min"}, {"max"}):
            raise PolicyValidationError(
                f"Gate '{metric}' must contain exactly one of 'min' or 'max'."
            )
        operator: GateOperator = "min" if "min" in raw_gate else "max"
        gates.append(
            PolicyGate(
                metric=metric,
                operator=operator,
                threshold=_as_number(raw_gate[operator], f"Gate '{metric}' threshold"),
            )
        )
    return ReleasePolicy(gates=tuple(gates))


def evaluate_policy(policy: ReleasePolicy, metrics: Mapping[str, Any]) -> PolicyEvaluation:
    """Evaluate policy gates in input order without external side effects."""
    results: list[GateEvaluation] = []
    for gate in policy.gates:
        if gate.metric not in metrics:
            results.append(
                GateEvaluation(
                    metric=gate.metric,
                    operator=gate.operator,
                    threshold=gate.threshold,
                    actual=None,
                    passed=False,
                    reason="Required metric was not supplied.",
                )
            )
            continue

        actual = _as_number(metrics[gate.metric], f"Metric '{gate.metric}'")
        passed = actual >= gate.threshold if gate.operator == "min" else actual <= gate.threshold
        results.append(
            GateEvaluation(
                metric=gate.metric,
                operator=gate.operator,
                threshold=gate.threshold,
                actual=actual,
                passed=passed,
            )
        )
    return PolicyEvaluation(passed=all(result.passed for result in results), results=tuple(results))

File: src/domain/policy.rs
//! Deterministic metric-gate parsing and evaluation.

use serde::{Deserialize, Serialize};
use serde_json::{Map, Value};

use super::errors::PolicyValidationError;

#[derive(Clone, Copy, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(rename_all = "lowercase")]
pub enum GateOperator {
    Min,
    Max,
}

impl GateOperator {
    #[must_use]
    pub const fn as_str(self) -> &'static str {
        match self {
            Self::Min => "min",
            Self::Max => "max",
        }
    }
}

#[derive(Clone, Debug, PartialEq)]
pub struct PolicyGate {
    pub metric: String,
    pub operator: GateOperator,
    pub threshold: f64,
}

#[derive(Clone, Debug, PartialEq)]
pub struct ReleasePolicy {
    pub gates: Vec<PolicyGate>,
}

#[derive(Clone, Debug, Deserialize, PartialEq, Serialize)]
pub struct GateEvaluation {
    pub metric: String,
    pub operator: GateOperator,
    pub threshold: f64,
    pub actual: Option<f64>,
    pub passed: bool,
    pub reason: Option<String>,
}

#[derive(Clone, Debug, Deserialize, PartialEq, Serialize)]
pub struct PolicyEvaluation {
    pub passed: bool,
    pub results: Vec<GateEvaluation>,
}

fn as_number(value: &Value, field_name: &str) -> Result<f64, PolicyValidationError> {
    let number = value
        .as_f64()
        .ok_or_else(|| PolicyValidationError(format!("{field_name} must be a number.")))?;
    if number.is_finite() {
        Ok(number)
    } else {
        Err(PolicyValidationError(format!(
            "{field_name} must be finite."
        )))
    }
}

/// # Errors
///
/// Returns [`PolicyValidationError`] when the policy does not have exactly one valid, non-empty
/// set of metric gates.
pub fn parse_policy(raw_policy: &Value) -> Result<ReleasePolicy, PolicyValidationError> {
    let policy = raw_policy
        .as_object()
        .ok_or_else(|| PolicyValidationError("Policy must be an object.".to_owned()))?;
    if policy.len() != 1 || !policy.contains_key("gates") {
        return Err(PolicyValidationError(
            "Policy must contain exactly one 'gates' field.".to_owned(),
        ));
    }
    let gates = policy["gates"]
        .as_object()
        .filter(|value| !value.is_empty())
        .ok_or_else(|| {
            PolicyValidationError("Policy gates must be a non-empty object.".to_owned())
        })?;
    gates
        .iter()
        .map(|(metric, raw_gate)| {
            if metric.trim().is_empty() {
                return Err(PolicyValidationError(
                    "Policy gate names must be non-empty strings.".to_owned(),
                ));
            }
            let gate = raw_gate.as_object().ok_or_else(|| {
                PolicyValidationError(format!("Gate '{metric}' must be an object."))
            })?;
            let (operator, raw_threshold) = match (gate.get("min"), gate.get("max"), gate.len()) {
                (Some(value), None, 1) => (GateOperator::Min, value),
                (None, Some(value), 1) => (GateOperator::Max, value),
                _ => {
                    return Err(PolicyValidationError(format!(
                        "Gate '{metric}' must contain exactly one of 'min' or 'max'."
                    )));
                }
            };
            Ok(PolicyGate {
                metric: metric.clone(),
                operator,
                threshold: as_number(raw_threshold, &format!("Gate '{metric}' threshold"))?,
            })
        })
        .collect::<Result<Vec<_>, _>>()
        .map(|gates| ReleasePolicy { gates })
}

/// # Errors
///
/// Returns [`PolicyValidationError`] when a supplied metric value is not a finite number.
pub fn evaluate_policy(
    policy: &ReleasePolicy,
    metrics: &Map<String, Value>,
) -> Result<PolicyEvaluation, PolicyValidationError> {
    let results = policy
        .gates
        .iter()
        .map(|gate| match metrics.get(&gate.metric) {
            None => Ok(GateEvaluation {
                metric: gate.metric.clone(),
                operator: gate.operator,
                threshold: gate.threshold,
                actual: None,
                passed: false,
                reason: Some("Required metric was not supplied.".to_owned()),
            }),
            Some(value) => {
                let actual = as_number(value, &format!("Metric '{}'", gate.metric))?;
                let passed = match gate.operator {
                    GateOperator::Min => actual >= gate.threshold,
                    GateOperator::Max => actual <= gate.threshold,
                };
                Ok(GateEvaluation {
                    metric: gate.metric.clone(),
                    operator: gate.operator,
                    threshold: gate.threshold,
                    actual: Some(actual),
                    passed,
                    reason: None,
                })
            }
        })
        .collect::<Result<Vec<_>, _>>()?;
    Ok(PolicyEvaluation {
        passed: results.iter().all(|result| result.passed),
        results,
    })
}
